### PR TITLE
Remove DB conversion from index.yaml

### LINKED ISSUE #13122: Remove support for the old YAML format in the spec database
As a developer I want to remove support for the deprecated YAML database format so that there will be less vestigial code to maintain.

### Rationale

In #2189 we changed the database format from YAML to JSON for performance reasons. To ease the transition for users the PR made it possible to read the old YAML format and output the new JSON one transparently. Since the change happened three years ago, it's probably safe to remove the code that takes care of this conversion in `database.py`.

### Description

The idea is to just go through the `database.py` module and remove any reference to the old database file and related code. Might also impact unit tests.

### Additional information
None

### PR DESCRIPTION
**Pull request by alalazo**
fixes #13122

Removed the code that was converting the old `index.yaml` format into `index.json`. Since the change happened in #2189 it should be considered safe to drop this (untested) code.

### CODE REVIEW

File: lib/spack/spack/database.py [hunk 1]
```diff
@@ -952,35 +924,24 @@ def _read(self):
...
-            if self.is_upstream:
-                raise UpstreamDatabaseLockingError(
-                    "No database index file is present, and upstream"
-                    " databases cannot generate an index file")
-            # The file doesn't exist, try to traverse the directory.
-            # reindex() takes its own write lock, so no lock here.
-            with lk.WriteTransaction(self.lock):
-                self._write(None, None, None)
-            self.reindex(spack.store.layout)
+        if self.is_upstream and not os.path.isfile(self._index_path):
```
Comment (scheibelp) [inline lib/spack/spack/database.py]: Based on the above changes, if `os.path.isfile(self._index_path) == True`, then we would have exited the function, so I don't think the extra `not os.path.isfile(self._index_path)` check is required here. IMO it's slightly clearer to keep this inside of the `else` clause, or you could do something like:

```
elif self.is_upstream:
  raise ...

self._write(...)
```

Comment (alalazo) [reply lib/spack/spack/database.py]: Done. Originally, I wrote that redundant  check on purpose to make clear to the reader that both conditions need to be satisfied to raise the exception. After re-reading the exception message I agree that it is clear enough to not cause confusion.

**Comment by alalazo** [comment]
@scheibelp Ready for another review! Thanks!

**Comment by scheibelp** [comment]
Thanks!

**Pull request merged**